app/helpers/advanced_messenger_helper.rb

```diff
@@ -26,58 +36,70 @@ def getUnreadNotificationsForIssue(issue)
         return [first_unread_notification_index, unread_notifications_for_current_user, unread_notifications_for_others]
     end
 
-    def getUnreadNotificationsForCurrentUserGroupByIssues() 
-        unread_notifications = Journal.where("read_by_users ILIKE ?", '%"' + User.current.id.to_s + '":{"read":0%')
-        unread_notifications.sort_by{|j| j.issue.id}
-        unread_notifications_per_issue = Hash.new
+    def getUnreadNotificationsGroupByIssues() 
+        unread_notifications = Journal.where("notes != '' AND notes IS NOT NULL AND read_by_users ILIKE ?", '%"' + User.current.id.to_s + '":{"read":0%')
+        getIssue = lambda {|entity| return entity.issue }
+        return getUnreadNotificationsGroupByParentEntity(unread_notifications, getIssue)
+    end
+
+    def getUnreadNotificationsGroupByTopic() 
+        getTopic = lambda {|entity| return entity.root }
+        unread_notifications = Message.where("read_by_users ILIKE ?", '%"' + User.current.id.to_s + '":{"read":0%')
+        return getUnreadNotificationsGroupByParentEntity(unread_notifications, getTopic);
+    end
+
+    def getUnreadNotificationsGroupByParentEntity(unread_notifications, getParent) 
+        grouped_unread_notifications = Hash.new
 
-        unread_notifications.each do |journal|
-            next if !journal.notes?
-            if unread_notifications_per_issue[journal.issue.id] != nil
-                unread_notification_status = unread_notifications_per_issue[journal.issue.id]
+        unread_notifications.each do |entity|
+            parentEntity = getParent.call(entity)
+            if grouped_unread_notifications[parentEntity.id] != nil
+                unread_notification_status = grouped_unread_notifications[parentEntity.id]
             else
                 # we set user to nil because is always the current user
-                unread_notification_status = UnreadNotificationsStatus.new(0, nil, journal.issue)
-                unread_notifications_per_issue[journal.issue.id] = unread_notification_status
+                unread_notification_status = UnreadNotificationsStatus.new(0, nil, parentEntity)
+                grouped_unread_notifications[parentEntity.id] = unread_notification_status
             end
             unread_notification_status.count += 1 
         end
-        return unread_notifications_per_issue
+        return grouped_unread_notifications
     end
 
     def getUnreadNotificationsForCurrentUserCount()
```

Comment:
I see you use this function a lot from js/templates. What happens if the User is not logged in? Crashes with nill?

I think we should add a check for user to be logged in here and not from where it is called. What do you think?

Reply:
This code is called only from myPage, and if the user is not logged in this page can not be accesed:
![image](https://github.com/famiprog/redmine_advanced_messenger/assets/4996054/294e2542-b1d0-41bf-86da-641fb154ba07)
So in case of no user logged in this code doesn't get called